kolibri/core/assets/src/api_resource.js

```diff
@@ -0,0 +1,263 @@
+  }
+  get url() {
+    return this.resource.modelUrl(this.id);
+  }
+  get id() {
+    return this.attributes[this.resource.idKey];
+  }
+  set(attributes) {
+    Object.assign(this.attributes, attributes);
+  }
+}
+
+/** Class representing a 'view' of a single API resource.
+ *  Contains different Model objects, depending on the parameters passed to its fetch method.
+ */
+class Collection {
+  constructor(data = [], resource) {
+    this.resource = resource;
+    if (!this.resource) {
+      throw new TypeError('resource must be defined');
+    }
+    this.models = [];
+    this._model_map = {};
+    this.synced = false;
+  }
+
+  /**
+   * Method to fetch data from the server for this collection.
+   * @param {object} params - an object of parameters to be parsed into GET parameters on the
+   * fetch.
+   * @returns {Promise} - Promise is resolved with Array of Model attributes when the XHR
+   * successfully returns, otherwise reject is called with the response object.
+   */
+  fetch({ params } = { params: {} }) {
+    return new Promise((resolve, reject) => {
+      // Do a fetch on the URL, with the parameters passed in.
+      client({ path: this.url, params }).then((response) => {
+        // Reset current models to only include ones from this fetch.
+        this.models = [];
+        this._model_map = {};
+        // Set response object - an Array - on the Collection to record the data.
+        this.set(response.entity);
+        // Mark that the fetch has completed.
+        this.synced = true;
+        // Return the data from the models, not the models themselves.
+        resolve(this.models.map((model) => model.attributes));
+      }, (response) => {
+        logging.error('An error occurred', response);
+        reject(response);
+      });
+    });
+  }
+  get url() {
+    return this.resource.collectionUrl;
+  }
+
+  /**
+   * Make a model a member of the collection - record in the models Array, and in the mapping
+   * from id to model. Will automatically instantiate Models for data passed in as objects, and
+   * deduplicate within the collection.
+   * @param {(Object|Model|Object[]|Model[])} models - Either an Array or single instance of an
+   * object or Model.
+   */
+  set(models) {
+    let modelsToSet;
+    if (!Array.isArray(models)) {
+      modelsToSet = [models];
+    } else {
+      modelsToSet = models;
+    }
+
+    modelsToSet.forEach((model) => {
+      // Note: this method ensures instantiation deduplication of models within the collection
+      //  and across collections.
+      const setModel = this.resource.addModel(model);
+      if (!this._model_map[model.id]) {
+        this._model_map[model.id] = setModel;
+        this.models.push(setModel);
+      }
+    });
+  }
+}
+
+/** Class representing a single API resource.
+ *  Contains references to all Models that have been fetched from the server.
+ *  Can also be subclassed in order to create custom behaviour for particular API resources.
+ */
```

Comment:
Just adding this as a note with an initial reaction, not having fully wrapped my head around this yet:

It seems a bit strange to me (at least coming from the Backbone model/collection system) to have Model and Collection be generic and then pass in a specific Resource instance, rather than the other way around (Resource as a factory for these, I guess).

Ah, but now that I read down further, I see that a Resource *is* in fact a factory of sorts, and people probably shouldn't be instantiating models/collections directly (perhaps that can be noted in the docstrings/pre-class-comments). Ah, in fact they cannot be used directly elsewhere, as the classes aren't exported.